apis/oam/v1alpha1/types.go

```diff
@@ -0,0 +1,663 @@
+)
+
+// A DefinitionReference refers to a CustomResourceDefinition by name.
+type DefinitionReference struct {
+	// Name of the referenced CustomResourceDefinition.
+	Name string `json:"name"`
+}
+
+// A WorkloadDefinitionSpec defines the desired state of a WorkloadDefinition.
+type WorkloadDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this workload
+	// kind.
+	Reference DefinitionReference `json:"definitionRef"`
+}
+
+// +kubebuilder:object:root=true
+
+// A WorkloadDefinition registers a kind of Kubernetes custom resource as a
+// valid OAM workload kind by referencing its CustomResourceDefinition. The CRD
+// is used to validate the schema of the workload when it is embedded in an OAM
+// Component.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type WorkloadDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec WorkloadDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// WorkloadDefinitionList contains a list of WorkloadDefinition.
+type WorkloadDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []WorkloadDefinition `json:"items"`
+}
+
+// A TraitDefinitionSpec defines the desired state of a TraitDefinition.
+type TraitDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this trait kind.
+	Reference DefinitionReference `json:"definitionRef"`
+
+	// AppliesToWorkloads specifies the list of workload kinds this trait
+	// applies to. Workload kinds are specified in kind.group/version format,
+	// e.g. server.core.oam.dev/v1alpha1. Traits that omit this field apply to
+	// all workload kinds.
+	// +optional
+	AppliesToWorkloads []string `json:"appliesToWorkloads,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// A TraitDefinition registers a kind of Kubernetes custom resource as a valid
+// OAM trait kind by referencing its CustomResourceDefinition. The CRD is used
+// to validate the schema of the trait when it is embedded in an OAM
+// ApplicationConfiguration.
+// +kubebuilder:printcolumn:JSONPath=".spec.definitionRef.name",name=DEFINITION-NAME,type=string
+type TraitDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec TraitDefinitionSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// TraitDefinitionList contains a list of TraitDefinition.
+type TraitDefinitionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []TraitDefinition `json:"items"`
+}
+
+// A ScopeDefinitionSpec defines the desired state of a ScopeDefinition.
+type ScopeDefinitionSpec struct {
+	// Reference to the CustomResourceDefinition that defines this scope kind.
+	Reference DefinitionReference `json:"definitionRef"`
```

Comment:
Shouldn't it be type of [ObjectReference(https://github.com/kubernetes/api/blob/9e5ffd1f1320950b238cfce291b926411f0af722/core/v1/generated.proto#L2144)?

Reply:
No. Quoting [API Conventions](https://github.com/kubernetes/community/blob/017dbe20/contributors/devel/sig-architecture/api-conventions.md#references-to-related-objects):

> References to specific objects, especially specific resource versions and/or specific fields of those objects, are specified using the ObjectReference type (or other types representing strict subsets of it).

The reasons we prefer to define our own types (that are strict subsets of ObjectReference) are explained in https://github.com/crossplaneio/crossplane-runtime/issues/49.